plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+                .withUserAgentPrefix("")
+                .withUserAgentSuffix("Trino-exchange");
+
+        this.pinS3ClientToCurrentRegion = config.isPinS3ClientToCurrentRegion();
+        this.endPoint = config.getS3Endpoint();
+        verify(!pinS3ClientToCurrentRegion || endPoint == null,
+                "Invalid configuration: either endpoint can be set or S3 client can be pinned to the current region");
+        this.s3 = createAmazonS3Client(clientConfig, config.getS3AwsAccessKey(), config.getS3AwsSecretKey());
+        this.streamingUploadPartSize = toIntExact(config.getS3StreamingPartSize().toBytes());
+    }
+
+    @Override
+    public void createDirectories(URI dir)
+            throws IOException
+    {
+        // no need to do anything for S3
+    }
+
+    @Override
+    public SliceInput getSliceInput(URI file, Optional<SecretKey> secretKey)
+            throws IOException
+    {
+        GetObjectRequest request = new GetObjectRequest(getBucketName(file), keyFromUri(file));
+        secretKey.map(SSECustomerKey::new).ifPresent(request::withSSECustomerKey);
+        return new InputStreamSliceInput(s3.getObject(request).getObjectContent());
+    }
+
+    @Override
+    public SliceOutput createSliceOutput(URI file, Optional<SecretKey> secretKey)
+    {
+        String bucketName = getBucketName(file);
+        String key = keyFromUri(file);
+
+        final Optional<SSECustomerKey> sseCustomerKey = secretKey.map(SSECustomerKey::new);
+
+        Supplier<String> uploadIdFactory = () -> initMultipartUpload(bucketName, key, sseCustomerKey).getUploadId();
+        return new S3ExchangeSliceOutput(s3, bucketName, key, uploadIdFactory, uploadExecutor, streamingUploadPartSize, sseCustomerKey);
+    }
+
+    private InitiateMultipartUploadResult initMultipartUpload(String bucket, String key, Optional<SSECustomerKey> sseCustomerKey)
+    {
+        InitiateMultipartUploadRequest request = new InitiateMultipartUploadRequest(bucket, key)
+                .withObjectMetadata(new ObjectMetadata())
+                .withCannedACL(CannedAccessControlList.Private);
+        sseCustomerKey.ifPresent(request::withSSECustomerKey);
+
+        return s3.initiateMultipartUpload(request);
+    }
+
+    @Override
+    public boolean exists(URI file)
+    {
+        return getS3ObjectMetadata(file) != null;
+    }
+
+    @Override
+    public void createFile(URI file)
+            throws IOException
+    {
+        s3.putObject(getBucketName(file), keyFromUri(file), "");
+    }
+
+    @Override
+    public void deleteRecursively(URI uri)
+            throws IOException
+    {
+        if (!isDirectory(uri)) {
+            s3.deleteObject(getBucketName(uri), keyFromUri(uri));
+        }
+        else {
+            listDirectories(uri).forEach(dir -> {
+                try {
+                    deleteRecursively(dir);
+                }
+                catch (IOException e) {
+                    e.printStackTrace();
```

Comment:
rethrow.
Or memoize and rethrow at the end of loop.

Reply:
Just realized that I don't need to catch and throw here.